README.md
## use

#### AListBuilder

```
  List<Map<String, String>> listALevelObj = [
    {'title': 'Vue', 'image': 'assets/Vue.png'},
    {'title': 'Web', 'image': 'assets/Web.png'},
    {'title': 'Flutter', 'image': 'assets/Flutter.png'},
    {'title': 'UnrealEngine', 'image': 'assets/UnrealEngine.png'}
  ];
  listA (){
    return AListBuilder(
      levelObj: listALevelObj,
      backgroundColor: const Color(0xFF2E2E48),
      onPressed: (selectedCategory) {
        debugPrint('click $selectedCategory ');
        setState(() {
          
        });
      },
    );
  }
```
![OneLevel](https://github.com/shAdow-XJY/list_twolevel/blob/master/example/assets/twolevellist2.jpg)

---

#### BListBuilder

```
List<Map<String, String>> listBLevelObj = [
    {'title':'title1'},
    {'title':'title2'},
    {'title':'title3'},
    {'title':'title4'},
    {'title':'title5'},
  ];
  listB (){
    return BListBuilder(
      levelObj: listBLevelObj,
      itemHeight: 50,
      backgroundColor: const Color(0xFF2E2E48),
      onPressed: (selectedTitle) {
        debugPrint('click $selectedTitle ');
        setState(() {
          
        });
      },
    );
  }
```
![TwoLevel](https://github.com/shAdow-XJY/list_twolevel/blob/master/example/assets/twolevellist1.jpg)

## Link

### pub link

[https://pub.dev/packages/list_twolevel](https://pub.dev/packages/list_twolevel)

### use in my blog

click to see:[shAdowPlusing](https://shadowplusing.website/#/)

### code source

Github:[shAdow-XJY/list_twolevel: a two level list builder --- flutter component. (github.com)](https://github.com/shAdow-XJY/list_twolevel)

Gitee:[list_twolevel: a two level list builder --- flutter component. (gitee.com)](https://gitee.com/shAdowPlusing/list_twolevel)

CHANGELOG.md
## 0.0.1

* a simple two level list builder. fill with Map<> Obj.

## 0.0.2

* solve text overflow

## 0.0.3

* separate into two type list builders, easy to use.

lib/src/list_A.dart
import 'package:flutter/material.dart';

class AListBuilder extends StatefulWidget {
  List<Map<String, String>> levelObj;
  Function(String selectedTitle)? onPressed;
  double? itemHeight;
  Color? backgroundColor;
  TextStyle? textStyle;
  EdgeInsetsGeometry? listPadding;
  EdgeInsetsGeometry? itemMargin;
  TextAlign? itemAlignment;
  AListBuilder({
    Key? key,
    required this.levelObj,
    this.onPressed,
    this.itemHeight = 100,
    this.backgroundColor,
    this.textStyle,
    this.listPadding,
    this.itemMargin,
    this.itemAlignment,
  }) : super(key: key);

  @override
  _AListBuilderState createState() => _AListBuilderState();
}

class _AListBuilderState extends State<AListBuilder> {
  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      padding: widget.listPadding??const EdgeInsets.symmetric(horizontal: 10.0),
      itemCount: widget.levelObj.length,
      itemBuilder: (context, index) => Container(
          height: widget.itemHeight,
          width: double.infinity,
          margin: widget.itemMargin??const EdgeInsets.only(bottom: 10, right: 10),
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(20),
            color: widget.backgroundColor ?? const Color(0xFF2E2E48),
            boxShadow: const [BoxShadow()],
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              SizedBox(
                height: 80,
                child: Padding(
                  padding: const EdgeInsets.all(16.0),
                  child: ClipRRect(
                    borderRadius: BorderRadius.circular(50),
                    child: Image.asset(widget.levelObj[index]['image'] ?? 'assets/default.png'),
                  ),
                ),
              ),
              Expanded(
                child: InkWell(
                  child: Text(
                    widget.levelObj[index]['title'] ?? '',
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                    textAlign: widget.itemAlignment??TextAlign.start,
                    style: widget.textStyle ??
                        TextStyle(
                            color: Colors.white,
                            fontSize: 35,
                            fontWeight: FontWeight.w900,
                            shadows: [
                              BoxShadow(
                                  blurRadius: 5,
                                  color:
                                  Colors.white.withOpacity(0.54)),
                            ]),
                  ),
                  onTap: () {
                    widget.onPressed!(widget.levelObj[index]['title'] ?? '');
                  },
                ),
              ),
              IconButton(
                icon: const Icon(Icons.keyboard_arrow_right),
                iconSize: 48,
                color: Colors.white,
                splashColor: Colors.transparent,
                highlightColor: Colors.transparent,
                onPressed: () {
                  widget.onPressed!(widget.levelObj[index]['title'] ?? '');
                },
              )
            ],
          )
      ),
    );
  }
}


lib/src/list_B.dart
import 'package:flutter/material.dart';

class BListBuilder extends StatefulWidget {
  List<Map<String, String>> levelObj;
  Function(String selectedTitle)? onPressed;
  double? itemHeight;
  Color? backgroundColor;
  TextStyle? textStyle;
  EdgeInsetsGeometry? listPadding;
  EdgeInsetsGeometry? itemMargin;
  TextAlign? itemAlignment;
  BListBuilder({
    Key? key,
    required this.levelObj,
    this.onPressed,
    this.itemHeight = 100,
    this.backgroundColor,
    this.textStyle,
    this.listPadding,
    this.itemMargin,
    this.itemAlignment,
  }) : super(key: key);

  @override
  _BListBuilderState createState() => _BListBuilderState();
}

class _BListBuilderState extends State<BListBuilder> {
  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      padding: widget.listPadding??const EdgeInsets.symmetric(horizontal: 10.0),
      itemCount: widget.levelObj.length,
      itemBuilder: (context, index) => Container(
          height: widget.itemHeight,
          width: double.infinity,
          margin: widget.itemMargin??const EdgeInsets.only(bottom: 10, right: 10),
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(20),
            color: widget.backgroundColor ?? const Color(0xFF2E2E48),
            boxShadow: const [BoxShadow()],
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const SizedBox(width: 40,),
              Expanded(
                child: InkWell(
                  child: Text(
                    widget.levelObj[index]['title'] ?? '',
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                    textAlign: widget.itemAlignment??TextAlign.center,
                    style: widget.textStyle ??
                        TextStyle(
                            color: Colors.white,
                            fontSize: 20,
                            fontWeight: FontWeight.w900,
                            shadows: [
                              BoxShadow(
                                  blurRadius: 5,
                                  color: Colors.white.withOpacity(0.54)),
                            ]),
                  ),
                  onTap: () {
                    widget.onPressed!(widget.levelObj[index]['title'] ?? '');
                  },
                ),
              ),
              const SizedBox(width: 40,),
            ],
          )
      ),
    );
  }
}


lib/src/twolevel_listbuider.dart
import 'package:flutter/material.dart';

class ListBuilder extends StatefulWidget {
  bool oneLevel;
  List<Map<String, String>> levelObj;
  Function(String selectedTitle)? onPressed;
  double? itemHeight;
  Color? backgroundColor;
  TextStyle? textStyle;
  EdgeInsetsGeometry? listPadding;
  EdgeInsetsGeometry? itemMargin;
  TextAlign? itemAlignment;
  ListBuilder({
    Key? key,
    required this.oneLevel,
    required this.levelObj,
    this.onPressed,
    this.itemHeight = 100,
    this.backgroundColor,
    this.textStyle,
    this.listPadding,
    this.itemMargin,
    this.itemAlignment,
  }) : super(key: key);

  @override
  _ListBuilderState createState() => _ListBuilderState();
}

class _ListBuilderState extends State<ListBuilder> {
  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      padding: widget.listPadding??const EdgeInsets.symmetric(horizontal: 10.0),
      itemCount: widget.levelObj.length,
      itemBuilder: (context, index) => Container(
          height: widget.itemHeight,
          width: double.infinity,
          margin: widget.itemMargin??const EdgeInsets.only(bottom: 10, right: 10),
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(20),
            color: widget.backgroundColor ?? const Color(0xFF2E2E48),
            boxShadow: const [BoxShadow()],
          ),
          child: widget.oneLevel
              ? Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    widget.levelObj[index]['image'] == null
                        ? const SizedBox(
                      height: 80,
                    )
                        : SizedBox(
                      height: 80,
                      child: Padding(
                        padding: const EdgeInsets.all(16.0),
                        child: ClipRRect(
                          borderRadius: BorderRadius.circular(50),
                          child: Image.asset(widget.levelObj[index]
                          ['image'] ??
                              'assets/default.png'),
                        ),
                      ),
                    ),
                    Expanded(
                      child: InkWell(
                        child: Text(
                          widget.levelObj[index]['title'] ?? 'default',
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          textAlign: widget.itemAlignment??TextAlign.start,
                          style: widget.textStyle ??
                              TextStyle(
                                  color: Colors.white,
                                  fontSize: 35,
                                  fontWeight: FontWeight.w900,
                                  shadows: [
                                    BoxShadow(
                                        blurRadius: 5,
                                        color:
                                        Colors.white.withOpacity(0.54)),
                                  ]),
                        ),
                        onTap: () {
                          widget.onPressed!(
                              widget.levelObj[index]['title'] ?? 'default');
                        },
                      ),
                    ),
                    IconButton(
                      icon: const Icon(Icons.keyboard_arrow_right),
                      iconSize: 48,
                      color: Colors.white,
                      splashColor: Colors.transparent,
                      highlightColor: Colors.transparent,
                      onPressed: () {
                        widget.onPressed!(
                            widget.levelObj[index]['title'] ?? 'default');
                      },
                    )
                  ],
                )
              : Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    const SizedBox(width: 40,),
                    Expanded(
                      child: InkWell(
                        child: Text(
                          widget.levelObj[index]['title'] ?? 'default',
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          textAlign: widget.itemAlignment??TextAlign.center,
                          style: widget.textStyle ??
                              TextStyle(
                                  color: Colors.white,
                                  fontSize: 20,
                                  fontWeight: FontWeight.w900,
                                  shadows: [
                                    BoxShadow(
                                        blurRadius: 5,
                                        color: Colors.white.withOpacity(0.54)),
                                  ]),
                        ),
                        onTap: () {
                          widget.onPressed!(
                              widget.levelObj[index]['title'] ?? 'default');
                        },
                      ),
                    ),
                    const SizedBox(width: 40,),
                  ],
                )),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:list_twolevel/list_A.dart';
import 'package:list_twolevel/list_B.dart';
void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Two Level List',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key,}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  bool click = true;

  List<Map<String, String>> listALevelObj = [
    {'title': 'Vue', 'image': 'assets/Vue.png'},
    {'title': 'Web', 'image': 'assets/Web.png'},
    {'title': 'Flutter', 'image': 'assets/Flutter.png'},
    {'title': 'UnrealEngine', 'image': 'assets/UnrealEngine.png'}
  ];
  listA (){
    return AListBuilder(
      levelObj: listALevelObj,
      backgroundColor: const Color(0xFF2E2E48),
      onPressed: (selectedCategory) {
        debugPrint('click $selectedCategory ');
        setState(() {
          click = !click;
        });
      },
    );
  }

  List<Map<String, String>> listBLevelObj = [
    {'title':'title1'},
    {'title':'title2'},
    {'title':'title3'},
    {'title':'title4'},
    {'title':'title5'},
    {'title':'title6'},
    {'title':'title7'},
    {'title':'title8'},
    {'title':'title9'},
    {'title':'title10'},
    {'title':'title11'},
    {'title':'title12'},
    {'title':'title13'},
    {'title':'title14'},
    {'title':'title15'},
  ];
  listB (){
    return BListBuilder(
      levelObj: listBLevelObj,
      itemHeight: 50,
      backgroundColor: const Color(0xFF2E2E48),
      onPressed: (selectedTitle) {
        debugPrint('click $selectedTitle ');
        setState(() {
          click = !click;
        });
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFF464667),
        body: click
            ? listA()
            : listB()
    );
  }
}
